mgr/volumes: Filter inherited snapshots while listing snapshots

kotreshhr opened the pull request:
Filter inherited snapshots resulted as part of a snapshot
at ancestor level while listing snapshots of a subvolume
and subvolumegroup

Fixes: https://tracker.ceph.com/issues/48501
Signed-off-by: Kotresh HR <khiremat@redhat.com>


<!--
Thank you for opening a pull request!  Here are some tips on creating
a well formatted contribution.

Please give your pull request a title like "[component]: [short description]"

This is the format for commit messages:

"""
[component]: [short description]

[A longer multiline description]

Fixes: [ticket URL on tracker.ceph.com, create one if necessary]
Signed-off-by: [Your Name] <[your email]>
"""

The Signed-off-by line is important, and it is your certification that
your contributions satisfy the Developers Certificate or Origin.  For
more detail, see SubmittingPatches.rst.

The component is the short name of a major daemon or subsystem,
something like "mon", "osd", "mds", "rbd, "rgw", etc. For ceph-mgr modules,
give the component as "mgr/<module name>" rather than a path into pybind.

For more examples, simply use "git log" and look at some historical commits.

This was just a quick overview.  More information for contributors is available here:
https://raw.githubusercontent.com/ceph/ceph/master/SubmittingPatches.rst

-->
## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

kotreshhr commented:
jenkins test docs

kotreshhr commented:
@ajarr @batrick @vshankar @ShyamsundarR 

With this PR, list_snapshots filters the inherited snapshosts which affects "fs subvolume rm" interface. 

The subvolume rm fails if any snapshots are present in the subvolume. The '--retain_snapshot' option needs to be used. What should be behavior in the case where the inherited snapshots are present in subvolume ? 

Until now, without this PR, subvolume rm fails even if inherited snapshots are present in subvolume. With this PR, subvolume rm succeeds if inherited snapshots are present.

kotreshhr commented:
jenkins test make check

kotreshhr commented:
jenkins test make check

ajarr commented on src/pybind/mgr/volumes/fs/fs_util.py:
```diff
@@ -75,6 +75,27 @@ def listdir(fs, dirpath):
+    """
+    Get the snap names from a given .snap directory path
+    """
+    if os.path.basename(snapdirpath) != b".snap":
```
snapshot directory base name '.snap' should not be hard coded? It is mgr.rados.conf_get('client_snapdir') or volspec.snapshot_prefix

kotreshhr replied:
Done

ajarr commented on src/pybind/mgr/volumes/fs/fs_util.py:
```diff
@@ -75,6 +75,27 @@ def listdir(fs, dirpath):
+            d = fs.readdir(dir_handle)
+            while d:
+                if (d.d_name not in (b".", b"..")) and d.is_dir():
+                    if not d.d_name.startswith(b"_"):
```
Do ancestor snapshots always begin with '_' or is the prefix configurable?

kotreshhr replied:
I don't think so. @batrick ?

batrick replied:
Always I believe.

ajarr replied:
maybe you could use inherited_snap(d.d_name) here?

ajarr reviewed (changes requested):
I did not know that user cannot create a fs snapshot with '\_' as first character in its name.
We should mention this in the docs.

ShyamsundarR commented:
> @ajarr @batrick @vshankar @ShyamsundarR
> 
> With this PR, list_snapshots filters the inherited snapshosts which affects "fs subvolume rm" interface.
> 
> The subvolume rm fails if any snapshots are present in the subvolume. The '--retain_snapshot' option needs to be used. What should be behavior in the case where the inherited snapshots are present in subvolume ?

The use case to consider would be restoring a subvolume from a group snapshot (which  is the parent level snapshot in this case). This would be a supported/desired operation in the future (in CSI) and the manner of restoring this would determine how the remove interface should behave..

For example, if we wanted to clone a subvolume within a subvolumegroup snapshot, how would that change the interface? Or, what interface would be provided for the same?

As an aside, if we state an entire subvolumegroup snapshot should be the only recovery granularity, then we need to fixup each of the subvolumes within the same, i.e update .meta of each subvolume within to reflect the current state, rather than retained state in the snapshot when we clone it.

I see that CSI is attempting to define both manners of restore, a single subvolume within a group or the entire group (for different use-cases).

If we can clone a subvolume from a subvolumegroup snapshot, we can support removing subvolumes that have parent snapshots (as is the case with this patch). I think this is where we would end up at, so the current patch behavior IMO is fine.

> 
> Until now, without this PR, subvolume rm fails even if inherited snapshots are present in subvolume. With this PR, subvolume rm succeeds if inherited snapshots are present.

kotreshhr commented:
> > @ajarr @batrick @vshankar @ShyamsundarR
> > With this PR, list_snapshots filters the inherited snapshosts which affects "fs subvolume rm" interface.
> > The subvolume rm fails if any snapshots are present in the subvolume. The '--retain_snapshot' option needs to be used. What should be behavior in the case where the inherited snapshots are present in subvolume ?
> 
> The use case to consider would be restoring a subvolume from a group snapshot (which is the parent level snapshot in this case). This would be a supported/desired operation in the future (in CSI) and the manner of restoring this would determine how the remove interface should behave..
> 
> For example, if we wanted to clone a subvolume within a subvolumegroup snapshot, how would that change the interface? Or, what interface would be provided for the same?
> 
> As an aside, if we state an entire subvolumegroup snapshot should be the only recovery granularity, then we need to fixup each of the subvolumes within the same, i.e update .meta of each subvolume within to reflect the current state, rather than retained state in the snapshot when we clone it.
> 
> I see that CSI is attempting to define both manners of restore, a single subvolume within a group or the entire group (for different use-cases).
> 
Single subvolume with in the group meaning from inherited snapshot right? If that's the case, losing inherited snapshot during subvolume rm is a problem right ?

> If we can clone a subvolume from a subvolumegroup snapshot, we can support removing subvolumes that have parent snapshots (as is the case with this patch). I think this is where we would end up at, so the current patch behavior IMO is fine.
> 
> > Until now, without this PR, subvolume rm fails even if inherited snapshots are present in subvolume. With this PR, subvolume rm succeeds if inherited snapshots are present.

ajarr commented:
> > > @ajarr @batrick @vshankar @ShyamsundarR
> > > With this PR, list_snapshots filters the inherited snapshosts which affects "fs subvolume rm" interface.
> > > The subvolume rm fails if any snapshots are present in the subvolume. The '--retain_snapshot' option needs to be used. What should be behavior in the case where the inherited snapshots are present in subvolume ?
> > 
> > 
> > The use case to consider would be restoring a subvolume from a group snapshot (which is the parent level snapshot in this case). This would be a supported/desired operation in the future (in CSI) and the manner of restoring this would determine how the remove interface should behave..
> > For example, if we wanted to clone a subvolume within a subvolumegroup snapshot, how would that change the interface? Or, what interface would be provided for the same?
> > As an aside, if we state an entire subvolumegroup snapshot should be the only recovery granularity, then we need to fixup each of the subvolumes within the same, i.e update .meta of each subvolume within to reflect the current state, rather than retained state in the snapshot when we clone it.
> > I see that CSI is attempting to define both manners of restore, a single subvolume within a group or the entire group (for different use-cases).
> 
> Single subvolume with in the group meaning from inherited snapshot right? If that's the case, losing inherited snapshot during subvolume rm is a problem right ?
> 

You can still access subvolume's snapshotted data from gp snapshot within,
<subvolgp-path>/.snap/<gp-snap-name>/<subvol>/<subvol-uuid>/

> > If we can clone a subvolume from a subvolumegroup snapshot, we can support removing subvolumes that have parent snapshots (as is the case with this patch). I think this is where we would end up at, so the current patch behavior IMO is fine.
> > > Until now, without this PR, subvolume rm fails even if inherited snapshots are present in subvolume. With this PR, subvolume rm succeeds if inherited snapshots are present.

ajarr commented on src/pybind/mgr/volumes/fs/operations/group.py:
```diff
@@ -81,7 +81,7 @@ def list_snapshots(self):
             dirpath = os.path.join(self.path,
                                    self.vol_spec.snapshot_dir_prefix.encode('utf-8'))
-            return listdir(self.fs, dirpath)
+            return listsnaps(self.fs, self.vol_spec, dirpath, filter_inherited_snaps=True)
```
maybe sufficient to pass snap_dir_prefix?
where snap_dir_prefix = self.vol_spec.snapshot_dir_prefix.encode('utf-8')

kotreshhr replied:
I thought so, but since this is an API, giving the control to user to pass both validation strings didn't sound correct to me.

ShyamsundarR commented:
> > > > @ajarr @batrick @vshankar @ShyamsundarR
> > > > With this PR, list_snapshots filters the inherited snapshosts which affects "fs subvolume rm" interface.
> > > > The subvolume rm fails if any snapshots are present in the subvolume. The '--retain_snapshot' option needs to be used. What should be behavior in the case where the inherited snapshots are present in subvolume ?
> > > 
> > > 
> > > The use case to consider would be restoring a subvolume from a group snapshot (which is the parent level snapshot in this case). This would be a supported/desired operation in the future (in CSI) and the manner of restoring this would determine how the remove interface should behave..
> > > For example, if we wanted to clone a subvolume within a subvolumegroup snapshot, how would that change the interface? Or, what interface would be provided for the same?
> > > As an aside, if we state an entire subvolumegroup snapshot should be the only recovery granularity, then we need to fixup each of the subvolumes within the same, i.e update .meta of each subvolume within to reflect the current state, rather than retained state in the snapshot when we clone it.
> > > I see that CSI is attempting to define both manners of restore, a single subvolume within a group or the entire group (for different use-cases).
> > 
> > 
> > Single subvolume with in the group meaning from inherited snapshot right? If that's the case, losing inherited snapshot during subvolume rm is a problem right ?
> 
> You can still access subvolume's snapshotted data from gp snapshot within,
> <subvolgp-path>/.snap/<gp-snap-name>/<subvol>/<subvol-uuid>/

I would assume the above. So, the clone a subvolume from group snapshot, would access the subvolume contents from the group snapshot as above. Which hence would not lose the inherited snapshot.

> 
> > > If we can clone a subvolume from a subvolumegroup snapshot, we can support removing subvolumes that have parent snapshots (as is the case with this patch). I think this is where we would end up at, so the current patch behavior IMO is fine.
> > > > Until now, without this PR, subvolume rm fails even if inherited snapshots are present in subvolume. With this PR, subvolume rm succeeds if inherited snapshots are present.

kotreshhr commented:
> > > > > @ajarr @batrick @vshankar @ShyamsundarR
> > > > > With this PR, list_snapshots filters the inherited snapshosts which affects "fs subvolume rm" interface.
> > > > > The subvolume rm fails if any snapshots are present in the subvolume. The '--retain_snapshot' option needs to be used. What should be behavior in the case where the inherited snapshots are present in subvolume ?
> > > > 
> > > > 
> > > > The use case to consider would be restoring a subvolume from a group snapshot (which is the parent level snapshot in this case). This would be a supported/desired operation in the future (in CSI) and the manner of restoring this would determine how the remove interface should behave..
> > > > For example, if we wanted to clone a subvolume within a subvolumegroup snapshot, how would that change the interface? Or, what interface would be provided for the same?
> > > > As an aside, if we state an entire subvolumegroup snapshot should be the only recovery granularity, then we need to fixup each of the subvolumes within the same, i.e update .meta of each subvolume within to reflect the current state, rather than retained state in the snapshot when we clone it.
> > > > I see that CSI is attempting to define both manners of restore, a single subvolume within a group or the entire group (for different use-cases).
> > > 
> > > 
> > > Single subvolume with in the group meaning from inherited snapshot right? If that's the case, losing inherited snapshot during subvolume rm is a problem right ?
> > 
> > 
> > You can still access subvolume's snapshotted data from gp snapshot within,
> > /.snap////
> 
> I would assume the above. So, the clone a subvolume from group snapshot, would access the subvolume contents from the group snapshot as above. Which hence would not lose the inherited snapshot.
> 

Yes, I verified. It's accessible. 

> > > > If we can clone a subvolume from a subvolumegroup snapshot, we can support removing subvolumes that have parent snapshots (as is the case with this patch). I think this is where we would end up at, so the current patch behavior IMO is fine.
> > > > > Until now, without this PR, subvolume rm fails even if inherited snapshots are present in subvolume. With this PR, subvolume rm succeeds if inherited snapshots are present.

ShyamsundarR reviewed (commented):
LGTM!

batrick commented on qa/tasks/cephfs/test_volumes.py:
```diff
@@ -1941,6 +1941,53 @@ def test_subvolume_snapshot_ls(self):
+        # Create snapshot at ancestral level
+        ancestral_snappath1 = os.path.join(".", "volumes", group, ".snap", "ancestral_snap_1")
+        ancestral_snappath2 = os.path.join(".", "volumes", group, ".snap", "ancestral_snap_2")
+        self.mount_a.run_shell(['mkdir', '-p', ancestral_snappath1])
```
```suggestion
        self.mount_a.run_shell(['mkdir', '-p', ancestral_snappath1, ancestral_snappath2])
```

batrick commented on qa/tasks/cephfs/test_volumes.py:
```diff
@@ -1941,6 +1941,53 @@ def test_subvolume_snapshot_ls(self):
+        ancestral_snappath1 = os.path.join(".", "volumes", group, ".snap", "ancestral_snap_1")
+        ancestral_snappath2 = os.path.join(".", "volumes", group, ".snap", "ancestral_snap_2")
+        self.mount_a.run_shell(['mkdir', '-p', ancestral_snappath1])
+        self.mount_a.run_shell(['mkdir', '-p', ancestral_snappath2])
```
```suggestion
```

batrick commented on qa/tasks/cephfs/test_volumes.py:
```diff
@@ -1941,6 +1941,53 @@ def test_subvolume_snapshot_ls(self):
+        self.mount_a.run_shell(['mkdir', '-p', ancestral_snappath2])
+
+        subvolsnapshotls = json.loads(self._fs_cmd('subvolume', 'snapshot', 'ls', self.volname, subvolume, group))
+        self.assertEqual(len(subvolsnapshotls), 3)
```
```suggestion
        self.assertEqual(len(subvolsnapshotls), snap_count)
```

batrick commented on qa/tasks/cephfs/test_volumes.py:
```diff
@@ -1941,6 +1941,53 @@ def test_subvolume_snapshot_ls(self):
+        self.assertEqual(len(subvolsnapshotls), 3)
+
+        # remove ancestral snapshots
+        self.mount_a.run_shell(['rmdir', ancestral_snappath1])
```
```suggestion
        self.mount_a.run_shell(['rmdir', ancestral_snappath1, ancestral_snappath2])
```

batrick commented on qa/tasks/cephfs/test_volumes.py:
```diff
@@ -1941,6 +1941,53 @@ def test_subvolume_snapshot_ls(self):
+
+        # remove ancestral snapshots
+        self.mount_a.run_shell(['rmdir', ancestral_snappath1])
+        self.mount_a.run_shell(['rmdir', ancestral_snappath2])
```
```suggestion
```

batrick commented on qa/tasks/cephfs/test_volumes.py:
```diff
@@ -1941,6 +1941,53 @@ def test_subvolume_snapshot_ls(self):
         # verify trash dir is clean
         self._wait_for_trash_empty()
 
+    def test_subvolume_inherited_snapshot_ls(self):
```
Need another test that snapshot info on an inherited snap fails.

Test you can't use `subvoluemm snapshot rm` on an inherited snap.

@ShyamsundarR does it make sense in CSI to have a subvolumegroup snapshot and subvolume snapshot with the same name? We can disambiguate them but it'd be better to not allow it to avoid confusion.

ShyamsundarR replied:
> @ShyamsundarR does it make sense in CSI to have a subvolumegroup snapshot and subvolume snapshot with the same name? We can disambiguate them but it'd be better to not allow it to avoid confusion.

The names for snapshots (group or per-subvolume) in CSI land is a UUID, so conflicts should not arise. The UUID is generated by the CSI code, so it is feasible to ensure that we do not reuse one that is is the superset of subvolume and group snapshots already present.

There is no requirement or need to use the same names, even if the user decides to name a group snapshot and individual subvolume snapshots the same. As there is a (actually two) layer of mapping between the user provided names and UUIDs that we generate.

IOW, CSI should be good even if there is a hard restriction on naming conflict.

batrick replied:
@kotreshhr  is this resolved?

kotreshhr replied:
@batrick You mean not allowing subvolumegroup and subvolume to be off same name ? I have not done anything w.r.t that.
I have added the tests for info and rm.

batrick replied:
> @batrick You mean not allowing subvolumegroup and subvolume to be off same name ? I have not done anything w.r.t that.

Yes. Let's prevent that in this patch.

kotreshhr replied:
We have removed subvolumegroup snapshot creation from mgr/volumes for now. So this PR avoids subvolume snapshot creation with same name as it's subvolumegroup snapshot name if exists (subvolumegroup snapshot created by other means). 

Since subvolumegroup snapshot creation is not supported, the converse check doesn't hold (i.e., creation of subvolumegroup snapshot with same name as subvolume snapshot name).

kotreshhr commented on src/pybind/mgr/volumes/fs/operations/versions/subvolume_v1.py:
```diff
@@ -362,6 +362,9 @@ def remove_snapshot(self, snapname):
 
     def snapshot_info(self, snapname):
+        if inherited_snap(snapname):
+            raise VolumeException(-errno.EINVAL,
```
Operations on inherited snaps return EINVAL(like snap create, rm), hence followed the same here though ENOENT also fits well here.

ajarr replied:
maybe it's safer to give error message that 'snapshot name is invalid'?

ajarr commented on src/pybind/mgr/volumes/fs/fs_util.py:
```diff
@@ -75,6 +75,35 @@ def listdir(fs, dirpath):
+    """
+    Returns True if the snapname is inherited else False
+    """
+    if snapname.startswith("_"):
```
```
return snapname.startswith("-)
```
should be sufficient

kotreshhr replied:
Done

ajarr commented on src/pybind/mgr/volumes/fs/fs_util.py:
```diff
@@ -75,6 +75,35 @@ def listdir(fs, dirpath):
+                if (d.d_name not in (b".", b"..")) and d.is_dir():
+                    if not d.d_name.startswith(b"_"):
+                        snaps.append(d.d_name)
+                    elif d.d_name.startswith(b"_") and not filter_inherited_snaps:
```
maybe use inherited_snap(d.d_name)?

kotreshhr replied:
Done

ajarr commented on qa/tasks/cephfs/test_volumes.py:
```diff
@@ -1941,6 +1941,153 @@ def test_subvolume_snapshot_ls(self):
+
+        # snapshot info on ancestral snapshot
+        try:
+            self._get_subvolume_snapshot_info(self.volname, subvolume, ancestral_snap_name, group)
```
what's the use of this check?

kotreshhr replied:
Added to validate both ancestor snapshot name and inherited snapshot name should fail with subvolume commands. Yeah it doesn't makes sense to check ancestral snapshot name as it's not present in subvolume name space at all. I will remove it.

ajarr commented on qa/tasks/cephfs/test_volumes.py:
```diff
@@ -1941,6 +1941,153 @@ def test_subvolume_snapshot_ls(self):
+        ancestral_snap_name = "ancestral_snap_1"
+        ancestral_snappath1 = os.path.join(".", "volumes", group, ".snap", ancestral_snap_name)
+        self.mount_a.run_shell(['mkdir', '-p', ancestral_snappath1])
+        inherited_snap = "_{0}_1099511627777".format(ancestral_snap_name)
```
1099511627777 how is the number determined?

kotreshhr replied:
Looks like snap realms inode number [1] @batrick ?


[1] https://github.com/ceph/ceph/blob/5fbae26e9eea55b88bb2eea479c336ea19d0ddeb/src/mds/mdstypes.cc#L973

batrick replied:
That's a test instance. The code should use the inode  number of the directory which was snapshoted.

batrick replied:
You should verify that this snapshot name exists for the subvolume by accessing it with the client mount. But that the snapshot name is invalid when accessed via the `subvolume snapshot info` command.

kotreshhr replied:
Done

ajarr commented on qa/tasks/cephfs/test_volumes.py:
```diff
@@ -1941,6 +1941,153 @@ def test_subvolume_snapshot_ls(self):
+
+        # ancester snapshot should not be deletable
+        try:
+            self._fs_cmd("subvolume", "snapshot", "rm", self.volname, subvolume, ancestral_snap_name, "--group_name", group)
```
not sure why this check is needed?

kotreshhr replied:
As said above, I will remove it.

ajarr commented on src/pybind/mgr/volumes/fs/fs_util.py:
```diff
@@ -75,6 +75,35 @@ def listdir(fs, dirpath):
         raise VolumeException(-e.args[0], e.args[1])
     return dirs
 
+def inherited_snap(snapname):
```
maybe name it `is_inherited_snap`

kotreshhr replied:
Done

ajarr commented:
@kotreshhr looks good. only remaining question, https://github.com/ceph/ceph/pull/38652#discussion_r551177452

batrick commented:
@kotreshhr please run through teuthology after addressing above comments.

kotreshhr commented:
jenkins test api

kotreshhr commented:
> @kotreshhr please run through teuthology after addressing above comments.

Addressed comments and triggered teuthology run

http://pulpito.front.sepia.ceph.com/khiremat-2021-01-13_12:52:33-fs-wip-khiremat-38652-filter-inherited-snap-distro-basic-smithi/
